internal/status.go

```diff
@@ -0,0 +1,158 @@
+	spinI                  int
+	indexCh                chan int
+	startTime              time.Time
+	sizingErr              error
+	ctx                    context.Context
+	mtx                    sync.RWMutex
+}
+
+var spinChars = [5]string{"-", "\\", "|", "/", "-"}
+
+const tickMs = 50
+
+// NewStatusLine creates and initializes a new StatusLine.
+func NewStatusLine(ctx context.Context, resources *[]Resource) *StatusLine {
+	st := StatusLine{}
+	st.resources = resources
+	st.indexCh = make(chan int)
+	st.ctx = ctx
+	return &st
+}
+
+// Increment informs the StatusLine that a resource (at index i in resource list) has finished downloading.
+func (st *StatusLine) Increment(i int) {
+	st.indexCh <- i
+}
+
+// Start begins the goroutine and loop that will update/print the status line.
+// Pass true to force SL to update (spinner and second counter) every 50ms.
+func (st *StatusLine) Start(doTick bool) {
+	st.startTime = time.Now()
+	go func() {
+		fmt.Print(st.GetStatusString())
+		for {
+			// Block until value is inserted into indexCh (>= 0 when resource finishes downloading, -1 every 50 milliseconds to keep timer and spinner updating).
+			var i int
+			select {
+			case i = <-st.indexCh:
+			case <-time.After(tickMs * time.Millisecond):
+				i = -1
+			case <-st.ctx.Done():
+				return
+			}
+			if i == -1 && !doTick {
+				continue
+			}
+
+			st.mtx.Lock()
+			if i != -1 {
+				st.numBytesDownloaded += st.resourceSizes[i]
+				st.numResourcesDownloaded++
+			}
+
+			// Update/rotate spinner.
+			st.spinI += 1
+			if st.spinI == len(spinChars) {
+				st.spinI = 0
+			}
+			st.mtx.Unlock()
+
+			fmt.Print(st.GetStatusString())
+			if st.numResourcesDownloaded == len(*st.resources) {
+				fmt.Println()
+				return
+			}
+
+		}
+	}()
+
+}
+
+// initResourceSizes fetches the size, in bytes, of each resource.
+func (st *StatusLine) InitResourcesSizes(timeoutMs int) error {
+	log.Debug().Msg("Fetching resource sizes")
+	st.resourceSizes = make([]int64, len(*st.resources))
+	for i := 0; i < len(st.resourceSizes); i++ {
+		st.resourceSizes[i] = 0
+	}
+
+	st.totalBytes = 0
+	for i, r := range *st.resources {
+		resource := r
+		httpClient := &http.Client{Timeout: time.Duration(timeoutMs) * time.Millisecond}
+		resp, err := httpClient.Head(resource.Urls[0])
+		if err != nil {
+			log.Debug().Msg("Error fetching resource sizes")
+			return err
+		}
+		st.totalBytes += resp.ContentLength
+		st.resourceSizes[i] = resp.ContentLength
+	}
+
+	return nil
+}
+
+// GetStatusString composes and returns the status line string for printing.
+func (st *StatusLine) GetStatusString() string {
+	st.mtx.RLock()
+	defer st.mtx.RUnlock()
+
+	var spinner string
+	if st.numResourcesDownloaded < len(*st.resources) {
+		spinner = spinChars[st.spinI]
+	} else {
+		spinner = "✔"
+	}
+
+	barStr := "["
+	if st.sizingErr == nil {
+		barLength := 20
+		if st.totalBytes < 20 {
+			barLength = int(st.totalBytes)
+		}
+		squareSize := st.totalBytes / int64(barLength)
```

Comment:
This can break if `barLength` is zero. You need to deal gracefully with this case.